### PR TITLE
 added gestures

### PR DESCRIPTION
**Pull request by SolMendiola**
# :star: Feature

---

#### :pencil2: Description:

Added gestures to tip: Zoom (pinch), Zoom (double tap), Pan and Toggle visibility of buttons

---

### CODE REVIEW

File: lib/ui/home/home_screen.dart [hunk 1]
```diff
@@ -1,50 +1,68 @@
...
+        onDrawerChanged: (stateDrawer) => cubit.setHomeFabState(stateDrawer),
+        extendBody: true,
+        extendBodyBehindAppBar: true,
+        key: _scaffoldKey,
+        floatingActionButtonLocation: FloatingActionButtonLocation.startFloat,
+        floatingActionButton: _buildFab(),
+        backgroundColor: context.theme.colors.background,
+        body: SafeArea(child: child),
+        drawer: AppDrawer(
+          tabsController: AutoTabsRouter.of(context),
```
Comment (fedecor9) [inline lib/ui/home/home_screen.dart:50]: Shouldn’t you use the tabsController defined in the builder function?

File: lib/ui/home/home_cubit.dart [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  final GeneralErrorHandler _errorHandler;
+
+  HomeCubit(this._errorHandler) : super(const HomeBaseState.state());
+
+  void setHomeFabState(bool stateDrawer) =>
+      emit(state.copyWith(hideFabMenu: stateDrawer));
+
+  void setToggleState() {
+    emit(state.copyWith(hideFabMenu: !state.hideFabMenu));
+    emit(state.copyWith(hideFavouriteFab: !state.hideFavouriteFab));
```
Comment (mirland) [inline lib/ui/home/home_cubit.dart:21]: you should emit only one value

File: lib/ui/home/home_cubit.dart [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+class HomeCubit extends Cubit<HomeBaseState> {
+  // ignore: unused_field
+  final GeneralErrorHandler _errorHandler;
+
+  HomeCubit(this._errorHandler) : super(const HomeBaseState.state());
+
+  void setHomeFabState(bool stateDrawer) =>
+      emit(state.copyWith(hideFabMenu: stateDrawer));
+
+  void setToggleState() {
```
Comment (mirland) [inline lib/ui/home/home_cubit.dart:19]: the method is not clear

File: lib/ui/tips/tips_screen.dart [hunk 4]
```diff
@@ -54,28 +58,40 @@ class _TipContentScreenState extends State<_TipContentScreen> {
...
-          controller: _pageController,
-          allowImplicitScrolling: true,
-          scrollDirection: Axis.vertical,
-          itemCount: state.tips.length,
-          onPageChanged: (index) => cubit
-            ..onTipDisplayed(state.tips[index])
-            ..setCurrentPage(index),
-          itemBuilder: (BuildContext context, int index) => DisplayTipWidget(
-            tip: state.tips[index],
+      builder: (context, state) => BlocBuilder<HomeCubit, HomeBaseState>(
```
Comment (mirland) [inline lib/ui/tips/tips_screen.dart:61]: it's not ok, you have just have the builder, use it

File: pubspec.yaml [hunk 5]
```diff
@@ -45,6 +45,7 @@ dependencies:
   sqflite: 2.2.0+3
   stack_trace: 1.10.0
   stock: 1.0.1
+  photo_view: ^0.14.0
```
Comment (mirland) [inline pubspec.yaml:48]: remove ^ and sort it

**Review by fedecor9 — APPROVED** [review]

File: lib/ui/common/custom_scaffold_fab.dart [hunk 6]
```diff
@@ -27,12 +29,15 @@ class MainScaffoldWithFab extends StatelessWidget {
         floatingActionButtonLocation: FloatingActionButtonLocation.endDocked,
         floatingActionButton: Padding(
           padding: const EdgeInsets.only(bottom: 15.0),
-          child: Fab(
-            state: state,
-            iconNotSelected: iconNotSelected,
-            action: action,
-            iconSelected: iconSelected,
-            border: border,
+          child: Visibility(
```
Comment (mirland) [inline lib/ui/common/custom_scaffold_fab.dart:32]: maybe you can add a fade animation

File: lib/ui/common/custom_scaffold_fab.dart [hunk 7]
```diff
@@ -27,12 +29,19 @@ class MainScaffoldWithFab extends StatelessWidget {
...
           padding: const EdgeInsets.only(bottom: 15.0),
-          child: Fab(
-            state: state,
-            iconNotSelected: iconNotSelected,
-            action: action,
-            iconSelected: iconSelected,
-            border: border,
+          child: AnimatedOpacity(
+            opacity: visibility! ? 1.0 : 0.0,
+            duration: const Duration(milliseconds: 10),
```
Comment (mirland) [inline lib/ui/common/custom_scaffold_fab.dart:34]: shouldn't be like 150?

File: lib/ui/favourites/list_favourites_tips_screen.dart [hunk 8]
```diff
@@ -17,33 +19,56 @@ class ListFavouritesTipsScreen extends StatelessWidget {
...
-                  top: 5,
-                  left: 10,
-                  right: 5,
-                  bottom: 5,
-                ),
-                shape: const CircleBorder(),
-                backgroundColor: context.theme.colors.primary.shade100,
-                alignment: Alignment.center,
+  Widget build(BuildContext context) => BlocProvider(
+        create: (context) => GlobalUICubit(),
```
Comment (mirland) [inline lib/ui/favourites/list_favourites_tips_screen.dart:23]: shouldn't you provide it in the section?

File: lib/ui/favourites/list_favourites_tips_screen.dart [hunk 9]
```diff
@@ -17,33 +19,56 @@ class ListFavouritesTipsScreen extends StatelessWidget {
...
+  final ShowTipsType showTipType;
+  final Tip? tip;
+
+  @override
+  Widget build(BuildContext context) =>
+      BlocBuilder<GlobalUICubit, GlobalUIState>(
+        builder: (context, state) => SafeArea(
+          child: Stack(
+            children: [
+              TipsScreen(
```
Comment (mirland) [inline lib/ui/favourites/list_favourites_tips_screen.dart:44]: shouldn't you use the cubit here? I mean, in the tip screen

File: lib/ui/home/home_screen.dart [hunk 10]
```diff
@@ -1,50 +1,79 @@
...
-        visible: !_hideFab,
-        child: Fab(
-          state: const FabState.notSelected(),
-          border: const CircleBorder(),
-          iconNotSelected: Icons.arrow_forward_ios,
-          action: () => _scaffoldKey.currentState!.openDrawer(),
+  Widget _buildFab() => BlocBuilder<GlobalUICubit, GlobalUIState>(
+        builder: (context, state) => AnimatedOpacity(
+          opacity: !state.hideFabMenu ? 1.0 : 0.0,
+          duration: const Duration(milliseconds: 10),
```
Comment (mirland) [inline lib/ui/home/home_screen.dart:67]: use a constant for the duration. And increase it

File: lib/ui/section/global_ui/global_ui_cubit.dart [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+part 'global_ui_state.dart';
+
+class GlobalUICubit extends Cubit<GlobalUIState> {
+  GlobalUICubit() : super(const GlobalUIState.state());
+
+  void toggleFabState() => emit(
+        state.copyWith(
+          hideFabMenu: !state.hideFabMenu,
+          hideFavouriteFab: !state.hideFavouriteFab,
```
Comment (mirland) [inline lib/ui/section/global_ui/global_ui_cubit.dart:14]: You cannot talk about favourite fabs ore menu, it's the global ui. Maybe just one that is named showUiActionComponents

File: lib/ui/tips/tips_screen.dart [hunk 12]
```diff
@@ -63,20 +74,28 @@ class _TipContentScreenState extends State<_TipContentScreen> {
...
+          backgroundDecoration:
+              BoxDecoration(color: context.theme.colors.background),
+          scrollPhysics: const BouncingScrollPhysics(),
+          builder: (BuildContext context, int index) =>
+              PhotoViewGalleryPageOptions(
+            imageProvider: NetworkImage(state.tips[index].imageUrl),
+            initialScale: PhotoViewComputedScale.contained,
+            minScale: PhotoViewComputedScale.contained,
+            maxScale: PhotoViewComputedScale.covered * 1.5,
+            onTapUp: (context, __, ___) => globalCubit.toggleFabState(),
```
Comment (mirland) [inline lib/ui/tips/tips_screen.dart:90]:
```suggestion
            onTapUp: (context, _, __) => globalCubit.toggleFabState(),
```

**Review by mirland — COMMENTED** [review]
I'd refactor the providers. I'd use the section. What do you think @f7deleon ?

**Review by f7deleon — APPROVED** [review]

**Review by mirland — APPROVED** [review]

**Pull request merged**